From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Allink <arclicious@vivaldi.net>
Date: Sun, 10 Jul 2022 10:12:04 +0100
Subject: [PATCH] Add Scissors configuration file


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 8621dba7f10ab822b5b99ce0d05da58823605cb6..9ba3115c075a52e790a2de65ae4118a877c7b684 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1968,6 +1968,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Scissors start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getScissorsConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Scissors end
+
         /**
          * Sends the component to the player
          *
