From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Allink <arclicious@vivaldi.net>
Date: Sun, 10 Jul 2022 10:12:04 +0100
Subject: [PATCH] Add Scissors configuration file


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 9f7029d327b55c8b1b9cfef8ea2678b87dad8ddb..070512a66b812984b6076bda67c5836468cc7251 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1967,6 +1967,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Scissors start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getScissorsConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Scissors end
+
         /**
          * Sends the component to the player
          *
