From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Allink <arclicious@vivaldi.net>
Date: Sun, 10 Jul 2022 10:12:04 +0100
Subject: [PATCH] Add Scissors configuration file


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index da5cab4246bd253fcc4e4d9574bdae1867ebb5ab..0f804fd9b9e0d069896e6be1509b35cabd445970 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1948,6 +1948,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Scissors start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getScissorsConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Scissors end
+
         /**
          * Sends the component to the player
          *
