From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Video <videogamesm12@gmail.com>
Date: Sun, 10 Dec 2023 13:11:22 -0600
Subject: [PATCH] UUID validation


diff --git a/src/main/java/net/minecraft/network/chat/HoverEvent.java b/src/main/java/net/minecraft/network/chat/HoverEvent.java
index 7fd85ae2ebd7225f06d874aa7e37fbdb89e3ea92..b2cc15da40aa7bbcfc234b4e147ed0e53f359efa 100644
--- a/src/main/java/net/minecraft/network/chat/HoverEvent.java
+++ b/src/main/java/net/minecraft/network/chat/HoverEvent.java
@@ -20,6 +20,7 @@ import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.TagParser;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.util.ExtraCodecs;
+import net.minecraft.util.GsonHelper;
 import net.minecraft.util.StringRepresentable;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.item.Item;
@@ -161,7 +162,14 @@ public class HoverEvent {
                 CompoundTag compoundTag = TagParser.parseTag(text.getString());
                 Component component = Component.Serializer.fromJson(compoundTag.getString("name"));
                 EntityType<?> entityType = BuiltInRegistries.ENTITY_TYPE.get(new ResourceLocation(compoundTag.getString("type")));
-                UUID uUID = UUID.fromString(compoundTag.getString("id"));
+                // Scissors start
+                UUID uUID;
+                try {
+                    uUID = UUID.fromString(compoundTag.getString("id"));
+                } catch (Exception ex) {
+                    return null;
+                }
+                // Scissors end
                 return DataResult.success(new HoverEvent.EntityTooltipInfo(entityType, uUID, component));
             } catch (Exception var5) {
                 return DataResult.error(() -> {
