From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Video <videogamesm12@gmail.com>
Date: Fri, 10 Jun 2022 23:26:03 -0500
Subject: [PATCH] Removes useless spammy error logging


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
index 4caf272f065a681bd7101d52fedf2642cc78d3bc..af5a1111a06789e1401f8c60bb4c8838587f2376 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
@@ -1798,8 +1798,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
                 resource = CraftNamespacedKey.fromMinecraft(key);
             }
         } catch (IllegalArgumentException ex) {
-            org.bukkit.Bukkit.getLogger().warning("Namespaced resource does not validate: " + key.toString());
-            ex.printStackTrace();
+            // Scissors - Don't log errors thrown by invalid namespaces when an error is thrown
         }
 
         return resource;
