From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Luna <lunahatesgogle@gmail.com>
Date: Sun, 13 Mar 2022 14:38:38 -0300
Subject: [PATCH] ItemEntity - Check if items are air before calling setItem


diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index c7f06c3cfb737bd17a706798bf9cf0e1af5f0cc0..4fd00227d035284688e641cb2c55a3214dcbaa38 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -445,11 +445,15 @@ public class ItemEntity extends Entity implements TraceableEntity {
 
         CompoundTag nbttagcompound1 = nbt.getCompound("Item");
 
-        this.setItem(ItemStack.of(nbttagcompound1));
-        if (this.getItem().isEmpty()) {
-            this.discard(null); // CraftBukkit - add Bukkit remove cause
+        // Scissors start
+        ItemStack item = ItemStack.of(nbttagcompound1);
+        if (item.isEmpty()) {
+            this.discard(null);
+            return;
         }
 
+        this.setItem(item);
+        // Scissors end
     }
 
     @Override
