From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Video <videogamesm12@gmail.com>
Date: Fri, 10 Jun 2022 23:26:03 -0500
Subject: [PATCH] Removes useless spammy error logging


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
index ac9fb7cf96ed36162c4f6c2a834b627d0120f3cc..5da793c0754a943b12ef01cac1072b71afd27b0e 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
@@ -1795,8 +1795,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
                 resource = CraftNamespacedKey.fromMinecraft(key);
             }
         } catch (IllegalArgumentException ex) {
-            org.bukkit.Bukkit.getLogger().warning("Namespaced resource does not validate: " + key.toString());
-            ex.printStackTrace();
+            // Scissors - Don't log errors thrown by invalid namespaces when an error is thrown
         }
 
         return resource;
