From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Telesphoreo <me@telesphoreo.me>
Date: Tue, 19 Apr 2022 18:27:04 -0500
Subject: [PATCH] fix this patch


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
index ef50d9747f2a027e356b05ad8d94ffb2fd4a423e..574afb19e43ec5b17ae2fbee7db36e3164fe0a0b 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
@@ -192,7 +192,7 @@ public abstract class AbstractMinecartContainer extends AbstractMinecart impleme
     @Override
     public void remove(Entity.RemovalReason reason) {
         if (!this.level.isClientSide && reason.shouldDestroy()) {
-            Containers.dropContents(this.level, (Entity) this, (Container) this);
+            Containers.dropContents(this.level, this, this);
         }
 
         super.remove(reason);
