From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Luna <lunahatesgogle@gmail.com>
Date: Sun, 13 Mar 2022 14:38:38 -0300
Subject: [PATCH] ItemEntity - Check if items are air before calling setItem


diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index 9ca37cbf829ff4240ae79fc283bcf9e7f2c728d2..bc71ca6e082a137d17ee0bf43de72095bd19ab3d 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -379,11 +379,14 @@ public class ItemEntity extends Entity {
 
         CompoundTag nbttagcompound1 = nbt.getCompound("Item");
 
-        this.setItem(ItemStack.of(nbttagcompound1));
+        // Scissors start
         if (this.getItem().isEmpty()) {
             this.discard();
+            return;
         }
 
+        this.setItem(ItemStack.of(nbttagcompound1));
+        // Scissors end
     }
 
     @Override
