From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Video <videogamesm12@gmail.com>
Date: Wed, 30 Mar 2022 02:01:55 -0600
Subject: [PATCH] Validate String UUIDs during the CompoundTag -> GameProfile
 conversion process


diff --git a/src/main/java/net/minecraft/nbt/NbtUtils.java b/src/main/java/net/minecraft/nbt/NbtUtils.java
index 1ccf1d22ac77b669fffd2a1bd7b1331cf0fa281a..6b31a32f9dac31b587693ed8001bf190786ceae8 100644
--- a/src/main/java/net/minecraft/nbt/NbtUtils.java
+++ b/src/main/java/net/minecraft/nbt/NbtUtils.java
@@ -81,7 +81,12 @@ public final class NbtUtils {
         }
         // Paper start - support string UUID's
         if (nbt.contains("Id", 8)) {
-            uUID = UUID.fromString(nbt.getString("Id"));
+            // Scissors start - Validate String UUIDs in game profiles
+            try {
+                uUID = UUID.fromString(nbt.getString("Id"));
+            } catch (Exception ignored) {
+            }
+            // Scissors end
         }
         // Paper end
 
