From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Luna <lunahatesgogle@gmail.com>
Date: Fri, 8 Apr 2022 23:38:12 -0300
Subject: [PATCH] Limit ListTags to 1024 elements


diff --git a/src/main/java/net/minecraft/nbt/ListTag.java b/src/main/java/net/minecraft/nbt/ListTag.java
index f346306d61854bdc6b395f5d8f799909537a1b50..fef39a78552188e90ebfcd608e7bf9e15c385bd1 100644
--- a/src/main/java/net/minecraft/nbt/ListTag.java
+++ b/src/main/java/net/minecraft/nbt/ListTag.java
@@ -30,6 +30,7 @@ public class ListTag extends CollectionTag<Tag> {
                         list.add(tagType.load(dataInput, i + 1, nbtAccounter));
                     }
 
+                    if(j > 1024) return new ListTag(Lists.newArrayListWithCapacity(0), b);
                     return new ListTag(list, b);
                 }
             }
