From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Telesphoreo <me@telesphoreo.me>
Date: Tue, 19 Apr 2022 20:08:56 -0500
Subject: [PATCH] Fix updater


diff --git a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
index 153c06bc00c4619012e77cb496ac57a7f2fc9409..400103cc5cfd6d796e1f2b1ab18e98aea2a0f858 100644
--- a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
@@ -22,7 +22,8 @@ import org.bukkit.Bukkit;
 
 public class PaperVersionFetcher implements VersionFetcher {
     private static final java.util.regex.Pattern VER_PATTERN = java.util.regex.Pattern.compile("^([0-9\\.]*)\\-.*R"); // R is an anchor, will always give '-R' at end
-    private static final String GITHUB_BRANCH_NAME = getBranch();
+    // Scissors - This must be changed manually for update checking to work
+    private static final String GITHUB_BRANCH_NAME = "ver/1.18.2";
     private static final String DOWNLOAD_PAGE = "https://ci.plex.us.org/job/Scissors/job/" + GITHUB_BRANCH_NAME;
     private static @Nullable String mcVer;
 
@@ -62,11 +63,6 @@ public class PaperVersionFetcher implements VersionFetcher {
         final var manifest = JarManifests.manifest(Bukkit.getServer().getClass());
         return manifest == null ? null : manifest.getMainAttributes().getValue("Git-Commit");
     }
-
-    private static String getBranch() {
-        final var manifest = JarManifests.manifest(Bukkit.getServer().getClass());
-        return manifest == null ? null : manifest.getMainAttributes().getValue("Git-Branch");
-    }
     // Scissors end
 
     private static Component getUpdateStatusMessage(@Nonnull String repo, @Nonnull String branch, @Nonnull String versionInfo) {
