From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Video <videogamesm12@gmail.com>
Date: Sat, 11 Jun 2022 23:05:12 -0500
Subject: [PATCH] Validate coordinates before attempting to get block entities
 when handling Creative Inventory packets


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 935c482be4bf3767e197a576d0e38eac51115dd4..ff31e7de543ee82e01b7d80382874732314504d6 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1843,6 +1843,18 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
     }
     // Spigot end
 
+    // Scissors start - Readd the following Paper code
+    // Paper start
+    private static final int SURVIVAL_PLACE_DISTANCE_SQUARED = 6 * 6;
+    private static final int CREATIVE_PLACE_DISTANCE_SQUARED = 7 * 7;
+    private boolean isOutsideOfReach(double x, double y, double z) {
+        Location eyeLoc = this.getCraftPlayer().getEyeLocation();
+        double reachDistance = org.bukkit.util.NumberConversions.square(eyeLoc.getX() - x) + org.bukkit.util.NumberConversions.square(eyeLoc.getY() - y) + org.bukkit.util.NumberConversions.square(eyeLoc.getZ() - z);
+        return reachDistance > (this.getCraftPlayer().getGameMode() == org.bukkit.GameMode.CREATIVE ? CREATIVE_PLACE_DISTANCE_SQUARED : SURVIVAL_PLACE_DISTANCE_SQUARED);
+    }
+    // Paper end
+    // Scissors end
+
     @Override
     public void handleUseItemOn(ServerboundUseItemOnPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.getLevel());
@@ -3188,15 +3200,22 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
             if (!itemstack.isEmpty() && nbttagcompound != null && nbttagcompound.contains("x") && nbttagcompound.contains("y") && nbttagcompound.contains("z") && this.player.getBukkitEntity().hasPermission("minecraft.nbt.copy")) { // Spigot
                 BlockPos blockposition = BlockEntity.getPosFromTag(nbttagcompound);
                 // Paper start
-                BlockEntity tileentity = null;
-                if (this.player.distanceToSqr(blockposition.getX(), blockposition.getY(), blockposition.getZ()) < 32 * 32 && this.player.getLevel().isLoadedAndInBounds(blockposition)) {
-                    tileentity = this.player.level.getBlockEntity(blockposition);
-                }
-                // Paper end
+                // Scissors start - Validate coordinates and whether the player can reach them
+                if (Level.isInSpawnableBounds(blockposition) && !isOutsideOfReach(blockposition.getX(), blockposition.getY(), blockposition.getZ()))
+                {
+                    BlockEntity tileentity = null;
+                    if (this.player.distanceToSqr(blockposition.getX(), blockposition.getY(), blockposition.getZ()) < 32 * 32 && this.player.getLevel().isLoadedAndInBounds(blockposition))
+                    {
+                        tileentity = this.player.level.getBlockEntity(blockposition);
+                    }
+                    // Paper end
 
-                if (tileentity != null) {
-                    tileentity.saveToItem(itemstack);
+                    if (tileentity != null)
+                    {
+                        tileentity.saveToItem(itemstack);
+                    }
                 }
+                // Scissors end
             }
 
             boolean flag1 = packet.getSlotNum() >= 1 && packet.getSlotNum() <= 45;
