From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Luna <lunahatesgogle@gmail.com>
Date: Mon, 11 Jul 2022 17:29:12 -0300
Subject: [PATCH] Validate block entity tag query positions


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index aaaa55f8574694e4b83824773a99461b2f36542a..24502eb6588ae1ea19d53b4af66c3ba78280de59 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1374,8 +1374,10 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
     @Override
     public void handleBlockEntityTagQuery(ServerboundBlockEntityTagQuery packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.serverLevel());
-        if (this.player.hasPermissions(2)) {
-            BlockEntity tileentity = this.player.level().getBlockEntity(packet.getPos());
+        // Scissors start - Validate block entity tag query positions
+        if (this.player.hasPermissions(2) && Level.isInSpawnableBounds(packet.getPos())) {
+            BlockEntity tileentity = this.player.level().isLoaded(packet.getPos()) ? this.player.level().getBlockEntity(packet.getPos()) : null;
+            // Scissors end
             CompoundTag nbttagcompound = tileentity != null ? tileentity.saveWithoutMetadata() : null;
 
             this.player.connection.send(new ClientboundTagQueryPacket(packet.getTransactionId(), nbttagcompound));
