From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Luna <lunahatesgogle@gmail.com>
Date: Fri, 28 Apr 2023 16:44:50 -0300
Subject: [PATCH] Prevent player banning using duplicate UUIDs


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 868951dc21aff541765b1f58f08cdf3c47446d25..f714b076dcaa78cb8a666ce2e9f24c66d6af0989 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1643,7 +1643,13 @@ public class ServerLevel extends Level implements WorldGenLevel {
         if (entity != null) {
             ServerLevel.LOGGER.warn("Force-added player with duplicate UUID {}", player.getUUID());
             entity.unRide();
-            this.removePlayerImmediately((ServerPlayer) entity, Entity.RemovalReason.DISCARDED);
+            // Scissors start - Prevent player banning using duplicate UUIDs
+            if (entity instanceof ServerPlayer serverPlayer) {
+                this.removePlayerImmediately(serverPlayer, Entity.RemovalReason.DISCARDED);
+            } else {
+                entity.discard();
+            }
+            // Scissors end
         }
 
         this.entityLookup.addNewEntity(player); // Paper - rewite chunk system
