From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Allink <arclicious@vivaldi.net>
Date: Mon, 20 Mar 2023 07:04:50 +0000
Subject: [PATCH] Patch invalid entity rotation log spam


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c19ea0dd30c39b1e7a30ab5bcc32820a5ee3d16a..4705a80cd759dce1fe5313417c7eb0e7be05e706 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4628,7 +4628,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
     public void setYRot(float yaw) {
         if (!Float.isFinite(yaw)) {
-            Util.logAndPauseIfInIde("Invalid entity rotation: " + yaw + ", discarding.");
+            // Scissors start - Patch invalid entity rotation log spam
+            //Util.logAndPauseIfInIde("Invalid entity rotation: " + yaw + ", discarding.");
+            // Scissors end
         } else {
             this.yRot = yaw;
         }
@@ -4640,7 +4642,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
     public void setXRot(float pitch) {
         if (!Float.isFinite(pitch)) {
-            Util.logAndPauseIfInIde("Invalid entity rotation: " + pitch + ", discarding.");
+            // Scissors start - Patch invalid entity rotation log spam
+            //Util.logAndPauseIfInIde("Invalid entity rotation: " + pitch + ", discarding.");
+            // Scissors end
         } else {
             this.xRot = pitch;
         }
