From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Allink <arclicious@vivaldi.net>
Date: Mon, 20 Mar 2023 07:04:50 +0000
Subject: [PATCH] Patch invalid entity rotation log spam


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 3bf0c06f5042e9ff5572714dfc6a641a01a65978..240d845ff72ec15657ff648a9b18b86eeb696215 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4670,7 +4670,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
     public void setXRot(float pitch) {
         if (!Float.isFinite(pitch)) {
-            Util.logAndPauseIfInIde("Invalid entity rotation: " + pitch + ", discarding.");
+            // Scissors - Patch invalid entity rotation log spam
         } else {
             this.xRot = pitch;
         }
