From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Allink <arclicious@vivaldi.net>
Date: Mon, 20 Mar 2023 07:04:50 +0000
Subject: [PATCH] Patch invalid entity rotation log spam


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index e3ac487e246875f5f9ee6f32746da33c0de8200e..4d7bdd8d87158948a3eb28473777eb939a1adac6 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4739,7 +4739,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
 
     public void setXRot(float pitch) {
         if (!Float.isFinite(pitch)) {
-            Util.logAndPauseIfInIde("Invalid entity rotation: " + pitch + ", discarding.");
+            // Scissors - Patch invalid entity rotation log spam
         } else {
             this.xRot = pitch;
         }
