From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Allink <arclicious@vivaldi.net>
Date: Mon, 20 Mar 2023 07:04:50 +0000
Subject: [PATCH] Patch invalid entity rotation log spam


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 234362618f40a91509deb1d4d7f537d86a68df77..127373f731fa3a15fe501a97887275166ea6ce30 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4481,7 +4481,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
     public void setYRot(float yaw) {
         if (!Float.isFinite(yaw)) {
-            Util.logAndPauseIfInIde("Invalid entity rotation: " + yaw + ", discarding.");
+            // Scissors start - Patch invalid entity rotation log spam
+            //Util.logAndPauseIfInIde("Invalid entity rotation: " + yaw + ", discarding.");
+            // Scissors end
         } else {
             this.yRot = yaw;
         }
@@ -4493,7 +4495,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
     public void setXRot(float pitch) {
         if (!Float.isFinite(pitch)) {
-            Util.logAndPauseIfInIde("Invalid entity rotation: " + pitch + ", discarding.");
+            // Scissors start - Patch invalid entity rotation log spam
+            //Util.logAndPauseIfInIde("Invalid entity rotation: " + pitch + ", discarding.");
+            // Scissors end
         } else {
             this.xRot = pitch;
         }
