From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Luna <lunahatesgogle@gmail.com>
Date: Fri, 28 Apr 2023 16:44:50 -0300
Subject: [PATCH] Prevent player banning using duplicate UUIDs


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index b78a9628a88f2a495ef6de74446a02a14d41a1f6..04f3d427dabf04b12c7fb6e7991c4af5e8d6e4a8 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1668,7 +1668,13 @@ public class ServerLevel extends Level implements WorldGenLevel {
         if (entity != null) {
             ServerLevel.LOGGER.warn("Force-added player with duplicate UUID {}", player.getUUID());
             entity.unRide();
-            this.removePlayerImmediately((ServerPlayer) entity, Entity.RemovalReason.DISCARDED);
+            // Scissors start - Prevent player banning using duplicate UUIDs
+            if (entity instanceof ServerPlayer serverPlayer) {
+                this.removePlayerImmediately(serverPlayer, Entity.RemovalReason.DISCARDED);
+            } else {
+                entity.discard();
+            }
+            // Scissors end
         }
 
         this.entityLookup.addNewEntity(player); // Paper - rewite chunk system
