From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Allink <arclicious@vivaldi.net>
Date: Mon, 20 Mar 2023 07:04:50 +0000
Subject: [PATCH] Patch invalid entity rotation log spam


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 67a5495c4048a40656b6b7b8ba2b9d1a78d88335..02c3a3aea71e7168f7953de0d39a9ea081dd8581 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4480,7 +4480,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
     public void setYRot(float yaw) {
         if (!Float.isFinite(yaw)) {
-            Util.logAndPauseIfInIde("Invalid entity rotation: " + yaw + ", discarding.");
+            // Scissors start - Patch invalid entity rotation log spam
+            //Util.logAndPauseIfInIde("Invalid entity rotation: " + yaw + ", discarding.");
+            // Scissors end
         } else {
             this.yRot = yaw;
         }
@@ -4492,7 +4494,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
     public void setXRot(float pitch) {
         if (!Float.isFinite(pitch)) {
-            Util.logAndPauseIfInIde("Invalid entity rotation: " + pitch + ", discarding.");
+            // Scissors start - Patch invalid entity rotation log spam
+            //Util.logAndPauseIfInIde("Invalid entity rotation: " + pitch + ", discarding.");
+            // Scissors end
         } else {
             this.xRot = pitch;
         }
