From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Luna <lunahatesgogle@gmail.com>
Date: Mon, 11 Jul 2022 17:10:17 -0300
Subject: [PATCH] Validate block entity tag query positions


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index a102586b42ac408a3f655f7c0d069ab0990ec7fc..3e6160650a2e22f9d4232d4656a176007b5bea9e 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1252,8 +1252,10 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
     @Override
     public void handleBlockEntityTagQuery(ServerboundBlockEntityTagQuery packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.getLevel());
-        if (this.player.hasPermissions(2)) {
-            BlockEntity tileentity = this.player.getLevel().getBlockEntity(packet.getPos());
+        // Scissors start - Validate block entity tag query positions
+        if (this.player.hasPermissions(2) && Level.isInSpawnableBounds(packet.getPos())) {
+            BlockEntity tileentity = this.player.getLevel().isLoaded(packet.getPos()) ? this.player.getLevel().getBlockEntity(packet.getPos()) : null;
+            // Scissors end
             CompoundTag nbttagcompound = tileentity != null ? tileentity.save(new CompoundTag()) : null;
 
             this.player.connection.send(new ClientboundTagQueryPacket(packet.getTransactionId(), nbttagcompound));
