From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Luna <lunahatesgogle@gmail.com>
Date: Fri, 28 Apr 2023 16:44:50 -0300
Subject: [PATCH] Prevent player banning using duplicate UUIDs


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 45804711255f04110e9509df8d60900314aa10b7..b1fd209b2893d2d6bfc2ae552c7919ab8abda695 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1468,7 +1468,14 @@ public class ServerLevel extends Level implements WorldGenLevel {
         if (entity != null) {
             ServerLevel.LOGGER.warn("Force-added player with duplicate UUID {}", player.getUUID().toString());
             entity.unRide();
-            this.removePlayerImmediately((ServerPlayer) entity, Entity.RemovalReason.DISCARDED);
+
+            // Scissors start - Prevent player banning using duplicate UUIDs
+            if (entity instanceof ServerPlayer serverPlayer) {
+                this.removePlayerImmediately(serverPlayer, Entity.RemovalReason.DISCARDED);
+            } else {
+                entity.discard();
+            }
+            // Scissors end
         }
 
         this.entityLookup.addNewEntity(player); // Paper - rewite chunk system
