From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Allink <arclicious@vivaldi.net>
Date: Wed, 5 Jul 2023 22:58:24 +0100
Subject: [PATCH] Limit save data for Bees and Vexes


diff --git a/src/main/java/net/minecraft/world/entity/animal/Bee.java b/src/main/java/net/minecraft/world/entity/animal/Bee.java
index f9521a6e115f0c975a7885b024c99eae300b63bf..c1598d0caee40a7d73af549b31b61ee7aae77315 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Bee.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Bee.java
@@ -234,8 +234,12 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
     @Override
     public void readAdditionalSaveData(CompoundTag nbt) {
         this.hivePos = null;
-        if (nbt.contains("HivePos")) {
-            this.hivePos = NbtUtils.readBlockPos(nbt.getCompound("HivePos"));
+        if (nbt.contains("HivePos"))
+        {
+            // Scissors start - Limit HivePos
+            final BlockPos savedHivePos = NbtUtils.readBlockPos(nbt.getCompound("HivePos"));
+            this.hivePos = this.level().isLoadedAndInBounds(savedHivePos) ? savedHivePos : null;
+            // Scissors end - Limit HivePos
         }
 
         this.savedFlowerPos = null;
diff --git a/src/main/java/net/minecraft/world/entity/monster/Vex.java b/src/main/java/net/minecraft/world/entity/monster/Vex.java
index f443006c1e32feee97b32312814e2447a50c45e2..ad418449f5ffd4bcceb5834b1385dc69259f3a21 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Vex.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Vex.java
@@ -118,8 +118,12 @@ public class Vex extends Monster implements TraceableEntity {
     @Override
     public void readAdditionalSaveData(CompoundTag nbt) {
         super.readAdditionalSaveData(nbt);
-        if (nbt.contains("BoundX")) {
-            this.boundOrigin = new BlockPos(nbt.getInt("BoundX"), nbt.getInt("BoundY"), nbt.getInt("BoundZ"));
+        if (nbt.contains("BoundX"))
+        {
+            // Scissors start - Limit Vex bound origin
+            final BlockPos savedBoundOrigin = new BlockPos(nbt.getInt("BoundX"), nbt.getInt("BoundY"), nbt.getInt("BoundZ"));
+            this.boundOrigin = this.level().isLoadedAndInBounds(savedBoundOrigin) ? savedBoundOrigin : null;
+            // Scissors end - Limit Vex bound origin
         }
 
         if (nbt.contains("LifeTicks")) {
